### DISCUSSION TITLE
Group Index Numbering

### DISCUSSION THREAD

**Discussion by Lucretiel** [rust-lang/regex#1188] [Q&A]
I'm working with `regex-syntax` on some stuff, and doing some transformations on the parsed syntax. Do I correctly understand that the `Capture` indices are assigned essentially lexically, from left to right? If I'm removing certain capture groups from the tree but keeping others, I can just renumber them in the order that they're encountered from left-to-right, starting at 1? I'm planning on using the `search_slots` API and want to make sure I'm doing everything correctly.

**Comment by BurntSushi** [comment]
Yes, it's done lexically. Each capture group index can be determined completely by just assigning an index (starting at `1`) to each opening parenthesis in the order in which it appears in the concrete syntax. So `(?<a>(?<b>(?<c>))|(?<d>))` has `a=1`, `b=2`, `c=3` and `d=4`. You can confirm this with `regex-cli`:

```
$ regex-cli debug hir '(?<a>(?<b>(?<c>))|(?<d>))'
    parse time:  24.225µs
translate time:  7.544µs

Capture(
    Capture {
        index: 1,
        name: Some(
            "a",
        ),
        sub: Alternation(
            [
                Capture(
                    Capture {
                        index: 2,
                        name: Some(
                            "b",
                        ),
                        sub: Capture(
                            Capture {
                                index: 3,
                                name: Some(
                                    "c",
                                ),
                                sub: Empty,
                            },
                        ),
                    },
                ),
                Capture(
                    Capture {
                        index: 4,
                        name: Some(
                            "d",
                        ),
                        sub: Empty,
                    },
                ),
            ],
        ),
    },
)
```
